### ISSUE TITLE
Batch processing doesn't respect the batch limit

### ISSUE THREAD

**Issue by Zelldon** [camunda/zeebe#11798]
**Describe the bug**

Discovered in https://zeebe-io.github.io/zeebe-chaos/2023/02/23/Recursive-call-activity/

Looks like that an recursive process can still make issues, even with the batch limiting feature.

![call](https://user-images.githubusercontent.com/2758593/220932389-a574970d-e26f-4e89-89d3-8ce957f85849.png)

**To Reproduce**

Run similar benchmark as  in https://zeebe-io.github.io/zeebe-chaos/2023/02/23/Recursive-call-activity/

**Expected behavior**

We process the instances in batches of 100 commands. There shouldn't be a noticeable impact like becoming unhealthy etc. We should be able to make progress on other places, other instances or can cancel the instances.

**Environment:**
- OS: 
- Zeebe Version: 8.2.0-SNAPSHOT, 8.1.9-SNAPSHOT 
- Configuration:

**Comment by Zelldon** [comment]
## Current investigation:

I went to an unit test and tried to investigate the issue with an debugger.

```
  @Test
  public void shouldActivateCallActivityForEver() throws InterruptedException {
    // when
    final var processInstanceKey =
        ENGINE.processInstance().ofBpmnProcessId(PROCESS_ID_PARENT).create();

    // then
    new CountDownLatch(1).await();
  }
```

So far I had no luck, it is right now not clear what the issue is.

## Back to metrics:**

I add a deeper look again at the metrics from yesterday.

![general](https://user-images.githubusercontent.com/2758593/221183514-f23fd431-4ec5-48af-beff-12884cd97023.png)

Based on the metrics it looks like that the processing dies after ~7 minutes.

![general-call](https://user-images.githubusercontent.com/2758593/221183557-e8a73e4f-496b-4275-855f-683d919437bc.png)

We can see that we successfully execute several call activities over minutes.

### Latency

The latency metrics show a draw picture. A) They show really high numbers ~136 years B) after around 1 minute they show that the latency increases.


![latency-batch](https://user-images.githubusercontent.com/2758593/221184206-a530537b-f3ec-4a32-babc-044a322c1fa3.png)

Interesting is that this is not only for the batch processing it is for all other, like commit latency and journal append etc.

![journal-latency](https://user-images.githubusercontent.com/2758593/221184201-11b2da9f-f099-41a9-960d-03261ac50375.png)
![commit-latency](https://user-images.githubusercontent.com/2758593/221184204-fb3ef124-c55e-4baf-a879-25d4008ffff3.png)

![flush-latency](https://user-images.githubusercontent.com/2758593/221184200-f1ae59e0-351a-41e2-93c8-41c93084687a.png)

With the sequencer we can see that the batchs are really huge.
![sequencer](https://user-images.githubusercontent.com/2758593/221184928-c2ad3cb2-fffe-494b-886d-be22f62960dc.png)

Based on the message metrics we see that they are chunked into 500 KB packages and send over the wire, not sure whether this is related or interesting.

![messaging](https://user-images.githubusercontent.com/2758593/221184924-5b5f1108-9c31-410d-a346-192ca0040d12.png)

**Comment by npepinpe** [comment]
I checked the following metric, `sum(increase(atomix_compaction_time_ms_bucket{namespace=~"np-flush-sync"}[5m])) by (le)`,  to compare in Grafana and in the Google Managed Prometheus console.

In Grafana, I get the following graph:

![image](https://user-images.githubusercontent.com/43373/221201253-dfc68229-5cf9-49f1-a166-3a1ea5c757df.png)

In Google Console, I get:

![image](https://user-images.githubusercontent.com/43373/221201354-3596f8c4-590a-40c2-b1d3-dac2350814be.png)

In the Prometheus console:

![image](https://user-images.githubusercontent.com/43373/221201855-9851ddfc-78c4-4cf6-a598-6621754b0690.png)

So to me it looks like it might be an issue in Grafana, but I wouldn't say this is 100% conclusive yet.

**Comment by Zelldon** [comment]
### Continued Investigation

We (@oleschoenburg @npepinpe @korthout ) were sitting in the TR and investigated this issue further.

TL;DR; this issue is related to RocksDB and not working well with bigger state. The recursive call activities are adding more and more data to the state, which eventually slows down the system completely.

#### Metrics

Something to add to @npepinpe comment, we have seen better latency metrics if we change the time frame, so there seems to be one metric report mess up all other.

#### Benchmark

We have run several benchmarks in order to understand what is going on. Thanks to @oleschoenburg suggestion and pre-work we investigated the benchmark with pyroscope.

```sh
helm install zell-chaos zeebe-benchmark/zeebe-benchmark --set starter.replicas=0 --set worker.replicas=0 --set profiling.enabled=true
$ k port-forward -n pyroscope pyroscope-5dc68d68fb-chvwm 4040
```

Deploying the process model and creating an instance:

```sh
$ zbchaos deploy process --processModelPath call.bpmn -v
Connecting to zell-chaos-again
Running experiment in self-managed environment.
Port forward to zell-chaos-zeebe-gateway-5f77c564df-mlzgh
Successfully created port forwarding tunnel
Deploy file call.bpmn (size: 2454 bytes).
Deployed process model call.bpmn successful with key 2251799813685249.
Deployed given process model call.bpmn, under key 2251799813685249!

$ zbchaos verify instance-creation --bpmnProcessId super -v
Connecting to zell-chaos-again
Running experiment in self-managed environment.
Port forward to zell-chaos-zeebe-gateway-5f77c564df-vx5qp
Successfully created port forwarding tunnel
Send create process instance command, with BPMN process ID 'super' and version '-1' (-1 means latest) [variables: '', awaitResult: false]
```

![diff_io camunda zeebe broker cpu_io camunda zeebe broker cpu_2023-02-24_1425-to-2023-02-24_1428](https://user-images.githubusercontent.com/2758593/221208303-a2093132-20da-46d1-902e-32d710cd8c24.png)

We have seen in the first flamegraphs, that when the throughput drop happens that the consistency checks are really taking long, so our assumption was that RocksDB might slow down on bigger state. This is something which we have seen already in the past.


We disabled the consistency check on a new benchmark:

```sh
$ helm install --create-namespace --namespace zell-chaos-again zell-chaos zeebe-benchmark/zeebe-benchmark --set starter.replicas=0 --set worker.replicas=0 --set zeebe.profiling.enabled=true --set zeebe.config.zeebe.broker.experimental.consistencyChecks.enablePreconditions=false --set zeebe.config.zeebe.broker.experimental.consistencyChecks.enableForeignKeyChecks=false
```

:warning: The zeebe benchmark are currently setting the checks to true, which makes it hard to set them to false. The statefulset had to be edit in order the remove the env vars. \cc @oleschoenburg 

When disabling the consistency checks the benchmark still failed after ~7 minutes. Based on the flamegraphs we can see that the blacklisting was also problematic.

![io camunda zeebe broker cpu_2023-02-24_1506-to-2023-02-24_1507](https://user-images.githubusercontent.com/2758593/221208448-788973d3-4d81-4d00-8914-c6d50bfb983e.png)

In order to understand whether Rocksdb is actually really the issue make, we increased the memory and run the same experiment.


```
$ helm install --create-namespace --namespace zell-chaos-again zell-chaos zeebe-benchmark/zeebe-benchmark --set starter.replicas=0 --set worker.replicas=0 --set zeebe.profiling.enabled=true --set zeebe.config.zeebe.broker.experimental.consistencyChecks.enablePreconditions=false --set zeebe.config.zeebe.broker.experimental.consistencyChecks.enableForeignKeyChecks=false --set zeebe.config.zeebe.broker.experimental.rocksdb.memoryLimit=1024MB
```

We can see in the image below, that before the throughput dropped imense after ~7 minutes. With the memory increase it continued to work much longer, and the drop after more than 15 minutes where not so big as before.

![works](https://user-images.githubusercontent.com/2758593/221208811-5c6e5d28-f028-4740-b7f4-b35463237745.png)

### Conclusion

What we can conclude out of this is that this is not an issue in the batch processing implementation itself. It is that we or our configured RocksDB instance can't handle big data properly. It will slow down tremendously if a certain limit is reached.

**Comment by Zelldon** [comment]
With #11814 we were able also to show that the batch limit metrics were wrong because of an issue in Grafana, so there is no bug in batch processing. Additionally as concluded above the issue lies here in the RocksDB big state handling, related to #11813. I will close this then.
